src/bin/pg_autoctl/pgsql.c

```diff
@@ -901,6 +901,8 @@ pgsql_execute_with_params(PGSQL *pgsql, const char *sql, int paramCount,
 
 	PQclear(result);
 	clear_results(pgsql);
+	PQfinish(pgsql->connection);
+	pgsql->connection = NULL;
```

Comment:
In a couple places in the code we are handling transactions to sync local state file creation with transaction commit on the monitor. When the local activity fails, we ROLLBACK the transaction on the monitor. I think we need to track if an explicit transaction is being used in our PGSQL object and provide `pgsql_begin`, `pgsql_commit`, and `pgsql_rollback` functions, and then have the `PGfinish` call in `pgsql_execute_with_params` depend on whether an explicit transaction is in flight or not.